cores/esp8266/uart.c

```diff
@@ -283,42 +289,105 @@ uart_resize_rx_buffer(uart_t* uart, size_t new_size)
 size_t
 uart_get_rx_buffer_size(uart_t* uart)
 {
-    return uart && uart->rx_enabled? uart->rx_buffer->size: 0;
+    return uart && uart->rx_enabled ? uart->rx_buffer->size : 0;
 }
 
 
-void ICACHE_RAM_ATTR 
-uart_isr(void * arg)
+void ICACHE_RAM_ATTR
+uart_isr(void* arg)
 {
-    uart_t* uart = (uart_t*)arg;
-    uint32_t usis = USIS(uart->uart_nr);
-
-    if(uart == NULL || !uart->rx_enabled) 
+    // check to confirm we are being called
+    if (arg == s_uartInterruptContext)
     {
-        USIC(uart->uart_nr) = usis;
-        ETS_UART_INTR_DISABLE();
-        return;
+        // Interrupt handler is shared between UART0 & UART1
+        // so we need to check both
+        for (uint8_t uartNum = 0; uartNum < 2; uartNum++)
+        {
+            if (USIS(uartNum))
+            {
+                if (s_uartInterruptContext[uartNum].callback != NULL)
+                {
+                    s_uartInterruptContext[uartNum].callback(s_uartInterruptContext[uartNum].arg);
+                }
+                else
+                {
+                    // No registered handler, so
+                    // clear all interrupts flags (just in case)
+                    USIE(uartNum) = 0;
+                    USIC(uartNum) = 0xffff;
+                }
+            }
+        }
     }
+}
 
-    if(usis & (1 << UIFF))
+void ICACHE_RAM_ATTR
+uart_isrDefault(void * arg)
+{
+    uart_t* uart = (uart_t*)arg;
+    uint32_t usis = USIS(uart->uart_nr);
+
+    if (usis & (1 << UIFF))
         uart_rx_copy_fifo_to_buffer_unsafe(uart);
 
-    if((usis & (1 << UIOF)) && !uart->rx_overrun)
+    if ((usis & (1 << UIOF)) && !uart->rx_overrun)
     {
         uart->rx_overrun = true;
         //os_printf_plus(overrun_str);
     }
-    
+
     if (usis & ((1 << UIFR) | (1 << UIPE) | (1 << UITO)))
         uart->rx_error = true;
 
     USIC(uart->uart_nr) = usis;
 }
 
-static void 
+void
+uart_subscribeInterrupt(int uart_nr, uartInterruptHandler callback, void* param)
+{
+    s_uartInterruptContext[uart_nr].callback = callback;
+    s_uartInterruptContext[uart_nr].arg = param;
+
+    int other_nr = (uart_nr + 1) % 2;
+    if (s_uartInterruptContext[other_nr].callback == NULL)
+    {
+        ETS_UART_INTR_ATTACH(uart_isr, s_uartInterruptContext);
```

Comment:
Maybe some comments like
`// attach interrupt if not already attached by the other port (1-uart_nr)`

Reply:
will do